### ISSUE TITLE
L1 penalty set small weights to 0

### ISSUE THREAD

**Issue by coco90417** [keras-team/tf-keras#16]
Dear team,

I have noticed that L1 penalty does not strictly force weights to 0. I wonder what is your view on below tentative approach that forces small weights to strictly 0?

class SmallWeightsToZero(Constraint):
    def __call__(self, w):
        mask = tf.abs(w) < tf.keras.backend.epsilon()  
        w = tf.where(mask, tf.zeros_like(w), w)
        return w 

model.add(layers.Dense(units = 1, 
                          kernel_regularizer = regularizers.L1(l1=l1),
                          kernel_constraint = SmallWeightsToZero()))

**Comment by coco90417** [comment]
@sachinprasadhs, kindly follow up on this thread. Please let us know if there is anything we can do to help.

**Comment by tilakrayal** [comment]
@coco90417,
The L1 penalty in vanilla form doesn't guarantee weights to be exactly zero. While it encourages sparsity by shrinking coefficients towards zero, some small, non-zero values can persist. 

This behavior is a key characteristic of L1 regularization and has both advantages and disadvantages depending on the context. The weights not being exactly 0 is an optimisation artifact. If you do linear least squares with L1 regularisation and optimise it with gradient descent you will never reach exactly 0. 

Also notice that for L1, the gradient (of the L1 term) is either 1 or -1, except for when w=0. That means that L1-regularization will move any weight towards 0 with the same step size, regardless of the weight’s value. In contrast, you can see that the L2 gradient is linearly decreasing towards 0 as the weight goes towards 0. Therefore, L2-regularization will also move any weight towards 0, but it will take smaller and smaller steps as a weight approaches 0. Thank you!
